src/middlewared/middlewared/plugins/sysdataset.py

```diff
@@ -500,24 +501,39 @@ async def setup_datasets(self, pool, uuid):
                 except Exception:
                     self.logger.warning("Failed to replace dataset [%s].", dataset, exc_info=True)
             else:
-                update_props_dict = {k: {'value': v} for k, v in props.items()
-                                     if datasets_prop[dataset][k]['value'] != v}
+                update_props_dict = {
+                    k: {'value': v} for k, v in props.items()
+                    if datasets_prop[dataset][k]['value'] != v
+                }
                 if update_props_dict:
                     await self.middleware.call(
                         'zfs.dataset.update',
                         dataset,
                         {'properties': update_props_dict},
                     )
 
+            await self.middleware.run_in_thread(self.__create_relevant_paths, config.get('create_paths', []))
+
+    def __create_relevant_paths(self, create_paths):
+        for create_path_config in create_paths:
+            os.makedirs(create_path_config['path'], exist_ok=True)
+            if all(create_path_config[k] for k in ('uid', 'gid')) and (
```

Comment:
we need to simplify this `all(` if branch. Stat'ing files/paths is relatively cheap, but we should still do it as little as possible.
```
for cpath in create_paths:
    os.makedirs(cpath['path'], exist_ok=True)
    cpath_stat = os.stat(cpath['path'])
    if cpath_stat.st_uid != cpath['uid'] or cpath_stat.st_gid != path['gid']:
        os.chown(cpath['path'], cpath['uid'], cpath['gid'])
```
Finally, since this is in our system dataset plugin, it is imperative that we wrap all these file I/O operations in a `try: except` clause so that we don't crash here. If system dataset fails, then it will cause a cascading set of failures for other services.